torch/distributed/_spmd/data_parallel.py

```diff
@@ -0,0 +1,763 @@
+            operand = node.all_input_nodes[0]
+            assert (
+                operand in self.batch_dim_map
+            ), "input have full reduction but not in dim map!"
+            self.set_batch_dim(node, self.get_batch_dim(operand))
+        else:
+            for i, dim_size in enumerate(shape):
+                if dim_size == self.batch_dim_size:
+                    self.set_batch_dim(node, i)
+
+        node_batch_dim = self.get_batch_dim(node)
+        batch_dim_shard_spec = DTensorSpec(
+            mesh=mesh, placements=[Shard(node_batch_dim)]
+        )
+        if reduction_over_batch or input_full_reduction:
+            batch_dim_shard_spec.from_local = True
+
+        return batch_dim_shard_spec, reduction_over_batch
+
+
+def _gen_shard_strategy(
+    mesh: DeviceMesh, shard_dim: int, input_specs: Optional[List[DTensorSpec]] = None
+) -> PlacementStrategy:
+    """
+    util function to generate a shard strategy on shard_dim
+    """
+    return PlacementStrategy(
+        output_spec=DTensorSpec(mesh=mesh, placements=[Shard(shard_dim)]),
+        input_specs=input_specs,
+    )
+
+
+def _gen_replicate_strategy(
+    mesh: DeviceMesh, input_specs: Optional[List[DTensorSpec]] = None
+) -> PlacementStrategy:
+    """
+    util function to generate a replicate strategy
+    """
+    return PlacementStrategy(
+        output_spec=DTensorSpec(mesh=mesh, placements=[Replicate()]),
+        input_specs=input_specs,
+    )
+
+
+def _gen_partial_strategy(mesh: DeviceMesh) -> PlacementStrategy:
+    """
+    util function to generate a partial strategy
+    """
+    return PlacementStrategy(
+        output_spec=DTensorSpec(mesh=mesh, placements=[_Partial()]),
+    )
+
+
+def build_data_parallel_strategies(
+    train_step_graph: GraphModule,
+    num_params: int,
+    num_states: int,
+    mesh: DeviceMesh,
+    batch_dim: int = 0,
+) -> Dict[fx.Node, StrategyList]:
+    """
+    This function loop through the train step graph and build the
+    data parallel strategy for each fx Node
+    """
+    activation_idx = num_params + num_states
+    non_compute_ops = [
+        torch.ops._spmd.tag_grad.default,
+        aten.ones_like.default,
+        aten.t.default,
+        aten.view.default,
+        aten.reshape.default,
+        aten.detach.default,
+        aten.clone.default,
+    ]
+
+    dp_strategy_map = {}
+    batch_dim_analzer = BatchDimAnalyzer(batch_dim)
+    placeholder_idx = 0
+    num_param_grad = 0
+
+    # first we backward propagate to mark the param gradients sharding
+    # with tag_grad node helps and then delete the tag_grad nodes
+    for node in reversed(list(train_step_graph.graph.nodes)):
+        # find a param_grad node via the tagging
+        if node.target == torch.ops._spmd.tag_grad.default:
+            cur_node = node
+            while cur_node.target in non_compute_ops:
+                cur_node = cur_node.args[0]
+                partial_strategy = _gen_partial_strategy(mesh)
```

Comment:
nit: do we expect mutations on `partial_strategy`? If not, this can be shared across all grad nodes?

Reply:
I was curious about whether we can treat `partial_strategy` as a constant? If yes, we only need one of that instead of creating a new one for every node?

Reply:
I see, good point! Let me think about this, maybe `replicate` and `partial` strategy can both be shared, only shard spec need to be generated separately

Reply:
If you are going to do this, I recommend turning the strategies into frozen dataclass, then you can singleton-ify them and the freeze will prevent people from trying to mutate the strategy.